Source/WebCore/rendering/RenderLayerBacking.cpp

```diff
@@ -2981,6 +3017,51 @@ bool RenderLayerBacking::paintsBoxDecorations() const
     return !supportsDirectlyCompositedBoxDecorations(renderer());
 }
 
+#if HAVE(HDR_SUPPORT)
+bool RenderLayerBacking::paintsHDRContent(RenderLayer::PaintedContentRequest& request) const
+{
+    auto& settings = renderer().settings();
+    if (!settings.hdrForImagesEnabled() && !settings.canvasPixelFormatEnabled())
+        return false;
+
+    auto hasPaintedHDRContent = [&settings = settings] (RenderElement& renderer) -> bool {
+#if ENABLE(HDR_FOR_IMAGES)
+        if (settings.hdrForImagesEnabled()) {
+            if (auto* imageRenderer = dynamicDowncast<RenderImage>(renderer)) {
+                if (auto* cachedImage = imageRenderer->cachedImage())
+                    return cachedImage->drawsHDRContent();
+                return false;
+            }
+        }
+#endif
+#if ENABLE(PIXEL_FORMAT_RGBA16F)
+        if (settings.canvasPixelFormatEnabled()) {
+            if (auto* canavsRenderer = dynamicDowncast<RenderHTMLCanvas>(renderer)) {
+                if (auto* renderingContext = canavsRenderer->canvasElement().renderingContext())
+                    return renderingContext->drawsHDRContent();
+                return false;
+            }
+        }
+#endif
+        return false;
+    };
+
+    if (hasPaintedHDRContent(renderer())) {
+        request.setHasPaintedHDRContent();
+        return true;
+    }
+
+    for (auto& child : descendantsOfType<RenderReplaced>(renderer())) {
```

Comment:
This code should be integrated with `RenderLayerBacking::paintsContent` so we only do one layer tree traversal;

Reply:
Fixed.